pyrit/datasets/fetch_examples.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import hashlib
+import io
+import tempfile
+from pathlib import Path
+
+import requests
+
+from pyrit.common.csv_helper import read_csv, write_csv
+from pyrit.common.json_helper import read_json, write_json
+from pyrit.common.txt_helper import read_txt, write_txt
+
+from typing import Callable, Dict, List, Optional, Literal, TextIO
+
+
+# Define the type for the file handlers
+FileHandlerRead = Callable[[TextIO], List[Dict[str, str]]]
+FileHandlerWrite = Callable[[TextIO, List[Dict[str, str]]], None]
+
+FILE_TYPE_HANDLERS: Dict[str, Dict[str, Callable]] = {
+    "json": {"read": read_json, "write": write_json},
+    "csv": {"read": read_csv, "write": write_csv},
+    "txt": {"read": read_txt, "write": write_txt},
+}
+
+
+def _get_cache_file_name(source: str, file_type: str) -> str:
+    """
+    Generate a cache file name based on the source URL and file type.
+    """
+    hash_source = hashlib.md5(source.encode("utf-8")).hexdigest()
+    return f"{hash_source}.{file_type}"
+
+
+def _read_cache(cache_file: Path, file_type: str) -> List[Dict[str, str]]:
+    """
+    Read data from cache.
+    """
+    with cache_file.open("r") as file:
+        if file_type in FILE_TYPE_HANDLERS:
+            return FILE_TYPE_HANDLERS[file_type]["read"](file)
+        else:
+            raise ValueError("Invalid file_type. Expected 'json', 'csv', or 'txt'.")
+
+
+def _write_cache(cache_file: Path, examples: List[Dict[str, str]], file_type: str):
+    """
+    Write data to cache.
+    """
+    cache_file.parent.mkdir(parents=True, exist_ok=True)
+    with cache_file.open("w") as file:
+        if file_type in FILE_TYPE_HANDLERS:
+            FILE_TYPE_HANDLERS[file_type]["write"](file, examples)
+        else:
+            raise ValueError("Invalid file_type. Expected 'json', 'csv', or 'txt'.")
+
+
+def _fetch_from_public_url(source: str, file_type: str) -> List[Dict[str, str]]:
+    """
+    Fetch examples from a repository.
+    """
+    response = requests.get(source)
+    if response.status_code == 200:
+        if file_type in FILE_TYPE_HANDLERS:
+            if file_type == "json":
+                return FILE_TYPE_HANDLERS[file_type]["read"](io.StringIO(response.text))
+            else:
+                return FILE_TYPE_HANDLERS[file_type]["read"](
+                    io.StringIO("\n".join(response.text.splitlines()))
+                )  # noqa: E501
+        else:
+            raise ValueError("Invalid file_type. Expected 'json', 'csv', or 'txt'.")
```

Comment:
When a new file_type is added, error message will be taken care automatically rather than updating the error message when a new file_type is added
```suggestion
            valid_types = ", ".join(FILE_TYPE_HANDLERS.keys())
            raise ValueError(f"Invalid file_type. Expected one of: {valid_types}.")
```